File: array-partition-i.js
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[]} nums
 * @return {number}
 */
var arrayPairSum = function(nums) {
    if (!nums || !nums.length) return 0;
    
    const buckets = new Array(20001).fill(0);
    const sortedNums = [];
    let sum = 0;
    for (const num of nums) {
        buckets[num + 10000]++;
    }

    for (let i = 0; i < buckets.length; i++) {
        if (buckets[i] == 0) continue;
        
        while(buckets[i] > 0) {
            sortedNums.push(i - 10000);
            buckets[i]--;
        }
    }

    for (let i = 0; i < sortedNums.length; i += 2) {
        sum += sortedNums[i];
    }
    return sum;
};

File: ArrayPartitionI.java
// Time Complexity : O(NlogN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
public class ArrayPartitionI {
    class Solution {
        public int arrayPairSum(int[] nums) {
            Arrays.sort(nums);
            int output = 0;
            for(int i =0; i < nums.length - 1; i+=2) {
                output += Math.min(nums[i], nums[i+1]);
            }
            return output;
        }
    }
}
